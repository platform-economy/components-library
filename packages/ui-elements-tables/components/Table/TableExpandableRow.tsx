import React, { Component, Fragment, Children } from 'react';
import styled from 'styled-components';
import { isUndefined } from 'lodash';
import classnames from 'classnames';

import { RenderContentWrapper } from './RenderContentWrapper';
import { TableCell } from './TableCell';

const StyledTableExpandableRow = styled.tr`
  &.expanded {
    background-color: ${({ theme }) => theme.palette.midground};
  }

  tbody > &.clickable:hover {
    background-color: ${({ theme }) => theme.palette.midground};
    cursor: pointer;
  }
`;

const StyledTableContentRow = styled.tr`
  @keyframes paddingTransition {
    from { padding: 0 30px; }
    to { padding: 18px 30px; }
  }

  overflow: hidden;
  height: min-content;

  ${TableCell} {
    animation: paddingTransition .5s;
  }

  &&.collapsed {
    border-top: none;

    ${TableCell} {
      padding: 0 30px;
    }
  }
`;

const StyledContentContainer = styled.div`
  transition: height .5s;
  overflow: hidden;
`;

export type TableExpandableRowProps = {
  expanded?: boolean;
  renderContent: (data: unknown) => React.ReactNode;
  renderData?: unknown;
  onClick?: (e: React.SyntheticEvent) => void;
};

export type TableExpandableRowState = {
  alreadyExpanded: boolean;
  contentHeight: number;
};

export class TableExpandableRow extends Component<
  TableExpandableRowProps, TableExpandableRowState
  > {
  constructor(props: TableExpandableRowProps) {
    super(props);

    this.state = {
      alreadyExpanded: !!props.expanded,
      contentHeight: 0,
    };
  }

  setHeight = (height: number) => {
    if (height !== this.state.contentHeight) {
      this.setState({
        contentHeight: height,
      });
    }
  }

  renderContent = () => {
    const { expanded, renderContent, renderData, children } = this.props;

    return (
      <StyledTableContentRow className={classnames({ collapsed: !expanded })}>
        <TableCell colSpan={Children.count(children)}>
          <StyledContentContainer
            style={{ height: expanded ? this.state.contentHeight : 0 }}
          >
            <RenderContentWrapper getHeight={this.setHeight}>
              {renderContent(renderData)}
            </RenderContentWrapper>
          </StyledContentContainer>
        </TableCell>
      </StyledTableContentRow>
    );
  }

  static getDerivedStateFromProps(props: TableExpandableRowProps, state: TableExpandableRowState) {
    if (!state.alreadyExpanded && props.expanded) {
      return { alreadyExpanded: true };
    }

    return null;
  }

  render() {
    const { children, expanded, onClick } = this.props;
    const { alreadyExpanded } = this.state;

    return (
      <Fragment>
        <StyledTableExpandableRow
          className={classnames({ clickable: !isUndefined(onClick), expanded: !!expanded })}
          onClick={onClick}
        >
          {children}
        </StyledTableExpandableRow>
        {alreadyExpanded && this.renderContent()}
      </Fragment>
    );
  }
}
