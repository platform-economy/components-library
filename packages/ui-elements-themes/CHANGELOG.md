# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.2](https://github.com/relayr/ui-elements/compare/@relayr/ui-elements-themes@0.1.1...@relayr/ui-elements-themes@0.1.2) (2019-08-22)

**Note:** Version bump only for package @relayr/ui-elements-themes






## [0.1.1](https://github.com/relayr/ui-elements/compare/@relayr/ui-elements-themes@0.1.0...@relayr/ui-elements-themes@0.1.1) (2019-08-21)

**Note:** Version bump only for package @relayr/ui-elements-themes






# 0.1.0 (2019-08-02)


### Features

* **ui-elements-basic:** new InputLabel component ([abb6076](https://github.com/relayr/ui-elements/commit/abb6076))
* **ui-elements-layout:** add Card component ([e8db0cf](https://github.com/relayr/ui-elements/commit/e8db0cf))
* **ui-elements-tables:** Table components ([#102](https://github.com/relayr/ui-elements/issues/102)) ([b1b8d6f](https://github.com/relayr/ui-elements/commit/b1b8d6f))
* **ui-elements-themes:** custom font linking system ([8e9f0f0](https://github.com/relayr/ui-elements/commit/8e9f0f0))
