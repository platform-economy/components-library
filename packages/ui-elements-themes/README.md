# ui-elements-themes

## Add to project

via npm:

```bash
npm install --save @relayr/ui-elements-themes
```

or Yarn:
```bash
yarn add @relayr/ui-elements-themes
```
