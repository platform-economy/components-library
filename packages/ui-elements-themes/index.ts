export * from './themes/light';
export * from './themes/dark';
export * from './themes/test';
export * from './themes/_common/paletteDark';
export * from './themes/_common/paletteLight';
export * from './themes/_common/methods';
export * from './themes/_common/fonts';
export * from './themes/_common/withTestTheme';
