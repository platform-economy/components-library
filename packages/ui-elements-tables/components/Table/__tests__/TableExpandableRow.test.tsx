import React, { ReactElement } from 'react';
import * as TestRenderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { configure, mount } from 'enzyme';

import 'jest-styled-components';

import { withTestTheme } from '@relayr/ui-elements-themes';

configure({ adapter: new Adapter() });

// Component
import { TableExpandableRow } from '../TableExpandableRow';

const renderFunction = jest.fn(() => (
  <p>Rendered content</p>
));

describe('ui-elements-tables', () => {
  describe('Components', () => {
    describe('TableExpandableRow', () => {
      it('Should render without throwing an exception', () => {
        const tree = TestRenderer
          .create(withTestTheme(
            <TableExpandableRow renderContent={renderFunction} />,
          ))
          .toJSON();

        expect(tree).toMatchSnapshot();
      });

      it('Should render with content', () => {
        const tree = TestRenderer
          .create(withTestTheme(
            <TableExpandableRow renderContent={renderFunction} expanded={true}>
              <td>Tabe data</td>
            </TableExpandableRow>,
          ))
          .toJSON();

        expect(tree).toMatchSnapshot();
      });

      it('Should handle onClick event', () => {
        const handler = jest.fn();
        const tree = mount(withTestTheme(
          <table>
            <tbody>
              <TableExpandableRow renderContent={renderFunction} onClick={handler} />
            </tbody>
          </table>,
        ));

        tree.find(TableExpandableRow).simulate('click');
        expect(handler).toHaveBeenCalledTimes(1);
      });

      it('Should handle onClick event without handler', () => {
        const tree = mount(withTestTheme(
          <table>
            <tbody>
              <TableExpandableRow renderContent={renderFunction} />
            </tbody>
          </table>,
        ));
        let error;

        try {
          tree.find(TableExpandableRow).simulate('click');
        } catch (err) {
          error = err;
        }

        expect(error).toBe(undefined);
      });

      it('Should update alreadyExpanded state value', () => {
        const WrappingComponent = (props: { children: ReactElement }) => (
          withTestTheme(
            <table>
              <tbody>
                {props.children}
              </tbody>
            </table>,
          )
        );
        const tree = mount(
          <TableExpandableRow renderContent={renderFunction} />,
          {
            wrappingComponent: WrappingComponent,
          },
        );
        const expectedStateBeforeUpdate = {
          alreadyExpanded: false,
          contentHeight: 0,
        };
        const expectedStateAfterUpdate = {
          alreadyExpanded: true,
          contentHeight: 0,
        };

        expect(tree.state()).toEqual(expectedStateBeforeUpdate);
        tree.setProps({ expanded: true });
        expect(tree.state()).toEqual(expectedStateAfterUpdate);
        tree.setProps({ expanded: false });
        expect(tree.state()).toEqual(expectedStateAfterUpdate);
        tree.setProps({ expanded: true });
        expect(tree.state()).toEqual(expectedStateAfterUpdate);
      });

      it('Should update height state value', () => {
        const WrappingComponent = (props: { children: ReactElement }) => (
          withTestTheme(
            <table>
              <tbody>
                {props.children}
              </tbody>
            </table>,
          )
        );
        const tree = mount<TableExpandableRow>(
          <TableExpandableRow renderContent={renderFunction} />,
          {
            wrappingComponent: WrappingComponent,
          },
        );
        const expectedStateBeforeUpdate = {
          alreadyExpanded: false,
          contentHeight: 0,
        };

        const expectedStateAfterUpdate = {
          alreadyExpanded: false,
          contentHeight: 10,
        };

        expect(tree.state()).toEqual(expectedStateBeforeUpdate);
        tree.instance().setHeight(10);
        expect(tree.state()).toEqual(expectedStateAfterUpdate);
      });
    });
  });
});
