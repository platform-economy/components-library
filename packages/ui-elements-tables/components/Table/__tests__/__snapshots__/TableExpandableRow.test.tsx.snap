// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui-elements-tables Components TableExpandableRow Should render with content 1`] = `
Array [
  .c0.expanded {
  background-color: #midground;
}

tbody > .c0.clickable:hover {
  background-color: #midground;
  cursor: pointer;
}

<tr
    className="expanded c0"
  >
    <td>
      Tabe data
    </td>
  </tr>,
  .c2 {
  -webkit-transition: padding .5s;
  transition: padding .5s;
  padding: 18px 16px;
  text-align: initial;
}

.c2:first-child {
  padding-left: 30px;
}

.c2:last-child {
  padding-right: 30px;
}

.c0 {
  overflow: hidden;
  height: -webkit-min-content;
  height: -moz-min-content;
  height: min-content;
}

.c0 .c1 {
  -webkit-animation: paddingTransition .5s;
  animation: paddingTransition .5s;
}

.c0.c0.collapsed {
  border-top: none;
}

.c0.c0.collapsed .c1 {
  padding: 0 30px;
}

.c3 {
  -webkit-transition: height .5s;
  transition: height .5s;
  overflow: hidden;
}

<tr
    className="c0"
  >
    <td
      className="c1 c2"
      colSpan={1}
    >
      <div
        className="c3"
        style={
          Object {
            "height": 0,
          }
        }
      >
        <div>
          <p>
            Rendered content
          </p>
        </div>
      </div>
    </td>
  </tr>,
]
`;

exports[`ui-elements-tables Components TableExpandableRow Should render without throwing an exception 1`] = `
.c0.expanded {
  background-color: #midground;
}

tbody > .c0.clickable:hover {
  background-color: #midground;
  cursor: pointer;
}

<tr
  className="c0"
/>
`;
